File: suc/swift/047.swift

import SwiftUI

// MARK: - AppTokens (Equivalent to Kotlin AppTokens object)
struct AppTokens {
    // MARK: Colors
    struct Colors {
        // Hex values for reference:
        // primary: #FF8C00
        // secondary: #FFC107
        // tertiary: #FFE082
        // background: #FFF8E1
        // surface: #FFFFFF
        // surfaceVariant: #FFECB3
        // outline: #D7CCC8
        // success: #4CAF50
        // warning: #FFB300
        // error: #E53935
        // onPrimary: #3E2723
        // onSecondary: #3E2723
        // onBackground: #3E2723
        // onSurface: #3E2723
        
        static let primary = Color(red: 255/255, green: 140/255, blue: 0/255)
        static let secondary = Color(red: 255/255, green: 193/255, blue: 7/255)
        static let tertiary = Color(red: 255/255, green: 224/255, blue: 130/255)
        static let background = Color(red: 255/255, green: 248/255, blue: 225/255)
        static let surface = Color(red: 255/255, green: 255/255, blue: 255/255)
        static let surfaceVariant = Color(red: 255/255, green: 236/255, blue: 179/255)
        static let outline = Color(red: 215/255, green: 204/255, blue: 200/255)
        static let success = Color(red: 76/255, green: 175/255, blue: 80/255)
        static let warning = Color(red: 255/255, green: 179/255, blue: 0/255)
        static let error = Color(red: 229/255, green: 57/255, blue: 53/255)
        static let onPrimary = Color(red: 62/255, green: 39/255, blue: 35/255)
        static let onSecondary = Color(red: 62/255, green: 39/255, blue: 35/255)
        static let onBackground = Color(red: 62/255, green: 39/255, blue: 35/255)
        static let onSurface = Color(red: 62/255, green: 39/255, blue: 35/255)
    }

    // MARK: TypographyTokens
    struct TypographyTokens {
        // Font sizes (sp) and weights
        static let display = Font.system(size: 28, weight: .bold)
        static let title = Font.system(size: 18, weight: .medium)
        static let body = Font.system(size: 14, weight: .regular) // Kotlin's FontWeight.Normal maps to .regular
        static let label = Font.system(size: 12, weight: .medium)
    }

    // MARK: Shapes
    struct Shapes {
        // Corner radii (dp)
        static let small = RoundedRectangle(cornerRadius: 6)
        static let medium = RoundedRectangle(cornerRadius: 10)
        static let large = RoundedRectangle(cornerRadius: 16)
    }

    // MARK: Spacing
    struct Spacing {
        // Spacing values (dp)
        static let sm: CGFloat = 6
        static let md: CGFloat = 10
        static let lg: CGFloat = 16
        static let xl: CGFloat = 24
    }

    // MARK: ShadowSpec (for elevation mapping)
    struct ShadowSpec {
        let elevation: CGFloat // Corresponds to blur radius in SwiftUI's .shadow
        let radius: CGFloat    // Not directly mapped as a separate blur radius in SwiftUI, elevation is used.
        let dy: CGFloat        // Y-offset for the shadow
        let opacity: Double    // Opacity of the shadow
    }

    // MARK: ElevationMapping
    struct ElevationMapping {
        static let level1 = ShadowSpec(elevation: 2, radius: 4, dy: 2, opacity: 0.1)
        static let level2 = ShadowSpec(elevation: 6, radius: 8, dy: 4, opacity: 0.15)
        static let level3 = ShadowSpec(elevation: 10, radius: 12, dy: 6, opacity: 0.18)
    }
}

// MARK: - Pin Data Model (Equivalent to Kotlin data class Pin)
struct Pin: Identifiable {
    let id = UUID() // Required for ForEach in SwiftUI to uniquely identify elements
    var x: CGFloat
    var y: CGFloat
    var label: String
}

// MARK: - RootScreen (Equivalent to Kotlin RootScreen Composable)
struct RootScreen: View {
    // @State for mutable state, equivalent to remember { mutableStateListOf(...) }
    @State private var pins: [Pin] = [
        Pin(x: 150, y: 250, label: "Group A"),
        Pin(x: 350, y: 420, label: "Group B"),
        Pin(x: 520, y: 310, label: "Group C")
    ]

    var body: some View {
        // Scaffold equivalent: A ZStack to set the background color for the entire screen,
        // and a VStack for the main content layout.
        ZStack {
            AppTokens.Colors.background // Set the background color
                .ignoresSafeArea(.all, edges: .all) // Make background fill under safe areas

            VStack(spacing: AppTokens.Spacing.lg) { // Column equivalent, with vertical arrangement spacedBy(lg)
                Text("Retro Group Buy")
                    .font(AppTokens.TypographyTokens.display)
                    .foregroundColor(AppTokens.Colors.onBackground)
                    .frame(maxWidth: .infinity, alignment: .leading) // Align text to the start (left)

                // Card equivalent for "Live Deals Map"
                VStack(alignment: .leading, spacing: AppTokens.Spacing.md) { // Column equivalent inside card
                    Text("Live Deals Map")
                        .font(AppTokens.TypographyTokens.title)
                        .foregroundColor(AppTokens.Colors.onSurface)

                    // Box equivalent for the map canvas
                    ZStack {
                        // Background for the map area, matching Kotlin's .background(AppTokens.Colors.surfaceVariant, AppTokens.Shapes.medium)
                        AppTokens.Colors.surfaceVariant
                            .clipShape(AppTokens.Shapes.medium) // Apply rounded corners

                        Canvas { context, size in
                            // Draw pins on the canvas
                            for pin in pins {
                                let center = CGPoint(x: pin.x, y: pin.y)
                                // drawCircle(radius = 14f) means diameter is 28f
                                let path = Path(ellipseIn: CGRect(x: center.x - 14, y: center.y - 14, width: 28, height: 28))
                                context.fill(path, with: .color(AppTokens.Colors.primary))
                            }
                        }
                        .padding(8) // Padding inside the Canvas, matching Kotlin's .padding(8.dp)
                    }
                    .frame(maxWidth: .infinity) // fillMaxWidth()
                    .frame(height: 220) // height(220.dp)
                    // The background and clipShape are primarily handled by the ZStack's first element,
                    // but adding them here again ensures the frame itself respects them if needed.
                    .background(AppTokens.Colors.surfaceVariant)
                    .clipShape(AppTokens.Shapes.medium)
                }
                .padding(AppTokens.Spacing.lg) // Padding inside the card, matching Kotlin's .padding(AppTokens.Spacing.lg)
                .background(AppTokens.Colors.surface) // Card background color
                .clipShape(AppTokens.Shapes.large) // Card shape
                .shadow(color: AppTokens.Colors.onSurface.opacity(AppTokens.ElevationMapping.level2.opacity),
                        radius: AppTokens.ElevationMapping.level2.elevation, // blur radius
                        x: 0, // No x-offset in Kotlin's shadow
                        y: AppTokens.ElevationMapping.level2.dy) // Card elevation

                Spacer().frame(height: AppTokens.Spacing.sm) // Spacer equivalent

                // Row equivalent for buttons
                HStack(spacing: AppTokens.Spacing.md) { // horizontalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md)
                    Button(action: {
                        // Random.nextInt(100, 550) generates 100 to 549
                        // Random.nextInt(150, 450) generates 150 to 449
                        let randomX = CGFloat.random(in: 100..<550)
                        let randomY = CGFloat.random(in: 150..<450)
                        pins.append(Pin(x: randomX, y: randomY, label: "New"))
                    }) {
                        Text("Add Group")
                            .font(AppTokens.TypographyTokens.label)
                            .foregroundColor(AppTokens.Colors.onSecondary)
                            .frame(maxWidth: .infinity) // Modifier.weight(1f) equivalent
                            .frame(height: 48) // height(48.dp)
                            .background(AppTokens.Colors.secondary) // containerColor
                            .clipShape(AppTokens.Shapes.medium) // shape
                    }
                    .buttonStyle(PlainButtonStyle()) // Remove default SwiftUI button styling for custom look

                    Button(action: {
                        pins.removeAll()
                    }) {
                        Text("Clear")
                            .font(AppTokens.TypographyTokens.label)
                            .foregroundColor(AppTokens.Colors.onPrimary)
                            .frame(maxWidth: .infinity) // Modifier.weight(1f) equivalent
                            .frame(height: 48) // height(48.dp)
                            .background(AppTokens.Colors.primary) // containerColor
                            .clipShape(AppTokens.Shapes.medium) // shape
                    }
                    .buttonStyle(PlainButtonStyle()) // Remove default SwiftUI button styling for custom look
                }

                // Card equivalent for "Active Groups"
                VStack(alignment: .leading, spacing: AppTokens.Spacing.sm) { // Column equivalent inside card
                    Text("Active Groups")
                        .font(AppTokens.TypographyTokens.title)
                        .foregroundColor(AppTokens.Colors.primary)

                    // ForEach for dynamic list items, equivalent to pins.forEach
                    ForEach(pins) { pin in
                        HStack { // Row equivalent
                            Text(pin.label)
                                .font(AppTokens.TypographyTokens.body)
                                .foregroundColor(AppTokens.Colors.onSurface)
                            Spacer() // horizontalArrangement = Arrangement.SpaceBetween
                            Circle() // Box with CircleShape
                                .fill(AppTokens.Colors.primary)
                                .frame(width: 14, height: 14) // size(14.dp)
                        }
                    }
                }
                .padding(AppTokens.Spacing.lg) // Padding inside the card
                .background(AppTokens.Colors.surface) // Card background color
                .clipShape(AppTokens.Shapes.large) // Card shape
                .shadow(color: AppTokens.Colors.onSurface.opacity(AppTokens.ElevationMapping.level1.opacity),
                        radius: AppTokens.ElevationMapping.level1.elevation, // blur radius
                        x: 0,
                        y: AppTokens.ElevationMapping.level1.dy) // Card elevation

                Spacer() // Pushes content to the top, similar to fillMaxSize() without specific alignment
            }
            .padding(AppTokens.Spacing.lg) // Outer padding for the whole column, matching .padding(pad).padding(AppTokens.Spacing.lg)
        }
        // MARK: Full-screen and Status Bar Hiding
        // .ignoresSafeArea() already used on the background color to extend content.
        // .statusBarHidden(true) hides the status bar.
        // For a more robust full-screen experience, also consider setting in Info.plist:
        // "View controller-based status bar appearance" to NO
        // "Status bar is initially hidden" to YES
        .statusBarHidden(true)
    }
}

// MARK: - App Entry Point (Equivalent to MainActivity)
@main
struct GroupBuyApp: App {
    var body: some Scene {
        WindowGroup {
            RootScreen()
        }
    }
}

File: suc/kotlin/047.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Shapes
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.Typography
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import kotlin.random.Random

private const val NAME = "061*GroupBuy*en"
private const val UI_TYPE = "ECommerce"
private const val STYLE_THEME = "Retro Flat"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFFFF8C00)
        val secondary = Color(0xFFFFC107)
        val tertiary = Color(0xFFFFE082)
        val background = Color(0xFFFFF8E1)
        val surface = Color(0xFFFFFFFF)
        val surfaceVariant = Color(0xFFFFECB3)
        val outline = Color(0xFFD7CCC8)
        val success = Color(0xFF4CAF50)
        val warning = Color(0xFFFFB300)
        val error = Color(0xFFE53935)
        val onPrimary = Color(0xFF3E2723)
        val onSecondary = Color(0xFF3E2723)
        val onBackground = Color(0xFF3E2723)
        val onSurface = Color(0xFF3E2723)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 28.sp, fontWeight = FontWeight.Bold)
        val title = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
    object Shapes {
        val small = RoundedCornerShape(6.dp)
        val medium = RoundedCornerShape(10.dp)
        val large = RoundedCornerShape(16.dp)
    }
    object Spacing {
        val sm = 6.dp
        val md = 10.dp
        val lg = 16.dp
        val xl = 24.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.1f)
        val level2 = ShadowSpec(6.dp, 8.dp, 4.dp, 0.15f)
        val level3 = ShadowSpec(10.dp, 12.dp, 6.dp, 0.18f)
    }
}

private val AppColorScheme = lightColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

data class Pin(val x: Float, val y: Float, val label: String)

@Composable
fun RootScreen() {
    val pins = remember {
        mutableStateListOf(
            Pin(150f, 250f, "Group A"),
            Pin(350f, 420f, "Group B"),
            Pin(520f, 310f, "Group C")
        )
    }
    Scaffold(
        contentWindowInsets = androidx.compose.foundation.layout.WindowInsets(0),
        containerColor = AppTokens.Colors.background
    ) { pad ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .padding(AppTokens.Spacing.lg),
            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.lg)
        ) {
            Text("Retro Group Buy", style = AppTokens.TypographyTokens.display, color = AppTokens.Colors.onBackground)
            Card(
                shape = AppTokens.Shapes.large,
                colors = CardDefaults.cardColors(containerColor = AppTokens.Colors.surface),
                elevation = CardDefaults.cardElevation(AppTokens.ElevationMapping.level2.elevation),
                modifier = Modifier.fillMaxWidth()
            ) {
                Column(
                    modifier = Modifier.fillMaxWidth().padding(AppTokens.Spacing.lg),
                    verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.md)
                ) {
                    Text("Live Deals Map", style = AppTokens.TypographyTokens.title, color = AppTokens.Colors.onSurface)
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(220.dp)
                            .background(AppTokens.Colors.surfaceVariant, AppTokens.Shapes.medium)
                    ) {
                        Canvas(modifier = Modifier.fillMaxSize().padding(8.dp)) {
                            drawRect(color = AppTokens.Colors.surfaceVariant)
                            pins.forEach {
                                drawCircle(color = AppTokens.Colors.primary, radius = 14f, center = Offset(it.x, it.y))
                            }
                        }
                    }
                }
            }
            Spacer(Modifier.height(AppTokens.Spacing.sm))
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Button(
                    onClick = {
                        pins.add(Pin(Random.nextInt(100, 550).toFloat(), Random.nextInt(150, 450).toFloat(), "New"))
                    },
                    colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.secondary, contentColor = AppTokens.Colors.onSecondary),
                    shape = AppTokens.Shapes.medium,
                    modifier = Modifier.weight(1f).height(48.dp)
                ) {
                    Text("Add Group", style = AppTokens.TypographyTokens.label)
                }
                Spacer(Modifier.width(AppTokens.Spacing.md))
                Button(
                    onClick = { pins.clear() },
                    colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.primary, contentColor = AppTokens.Colors.onPrimary),
                    shape = AppTokens.Shapes.medium,
                    modifier = Modifier.weight(1f).height(48.dp)
                ) {
                    Text("Clear", style = AppTokens.TypographyTokens.label)
                }
            }
            Card(
                shape = AppTokens.Shapes.large,
                colors = CardDefaults.cardColors(containerColor = AppTokens.Colors.surface),
                elevation = CardDefaults.cardElevation(AppTokens.ElevationMapping.level1.elevation),
                modifier = Modifier.fillMaxWidth()
            ) {
                Column(
                    modifier = Modifier.fillMaxWidth().padding(AppTokens.Spacing.lg),
                    verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.sm)
                ) {
                    Text("Active Groups", style = AppTokens.TypographyTokens.title, color = AppTokens.Colors.primary)
                    pins.forEach {
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.SpaceBetween,
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Text(it.label, style = AppTokens.TypographyTokens.body, color = AppTokens.Colors.onSurface)
                            Box(modifier = Modifier.size(14.dp).background(AppTokens.Colors.primary, CircleShape))
                        }
                    }
                }
            }
        }
    }
}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFFFF8E1)
@Composable
fun PreviewRoot() {
    AppTheme {
        RootScreen()
    }
}
